internal/agent/agent.go

```diff
@@ -0,0 +1,118 @@
+package agent
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/rombintu/goyametricsv2/internal/config"
+	"github.com/rombintu/goyametricsv2/internal/storage"
+)
+
+type Agent struct {
+	serverAddress  string
+	pollInterval   int64
+	reportInterval int64
+	data           map[string]interface{}
+	pollCount      int
+	metrics        map[string]string
+}
+
+func NewAgent(c config.AgentConfig) *Agent {
+	data := make(map[string]interface{})
+	data[storage.CounterType] = make(storage.CounterTable)
+	data[storage.GaugeType] = make(storage.GaugeTable)
+	return &Agent{
+		serverAddress:  fixServerURL(c.Address),
+		pollInterval:   c.PollInterval,
+		reportInterval: c.ReportInterval,
+		data:           data,
+		metrics:        make(map[string]string),
+	}
+}
+
+func fixServerURL(url string) string {
+	if strings.HasPrefix(url, "http://") {
+		return url
+	} else {
+		return fmt.Sprintf("http://%s", url)
+	}
+}
+
+func (a *Agent) incPollCount() {
+	a.pollCount++
+}
+
+func (a *Agent) postRequest(url string) error {
+	req, err := http.NewRequest(http.MethodPost, url, nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "text/plain")
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+	return nil
+}
+
+func (a *Agent) sendDataOnServer(metricType, metricName string, value string) error {
+	url := fmt.Sprintf("%s/update/%s/%s/%s", a.serverAddress, metricType, metricName, value)
+	if err := a.postRequest(url); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a *Agent) RunPoll() {
+	for {
+		a.loadMetrics()
+		time.Sleep(time.Duration(a.pollInterval) * time.Second)
+	}
+}
+
+func (a *Agent) RunReport() {
+	for {
+		for metricName, value := range a.metrics {
+			a.sendDataOnServer(storage.GaugeType, metricName, value)
+		}
+		a.sendDataOnServer(storage.CounterType, "pollCount", strconv.Itoa(a.pollCount))
+		time.Sleep(time.Duration(a.reportInterval) * time.Second)
```

Comment:
Тут по-моему можно использование `time.Duration(...)`  опустить.
```suggestion
		time.Sleep(a.reportInterval * time.Second)
```